README.md

```diff
@@ -368,63 +343,126 @@ fun main() {
+    val reconstructedPatient = json.decodeFromString<Resource>(jsonString)  // Deserialization
+    
+    check(reconstructedPatient is Patient)
+}
+```
+
+#### Polymorphism
+
+In the example above, notice the type parameter `<Resource>` in the serialization and
+deserialization function calls <code>encodeToString<b>\<Resource\></b></code> and
+<code>decodeFromString<b>\<Resource\></b></code>. This is a critical detail.
+
+To allow the `kotlinx.serialization` library to determine the resource type at runtime during
+_deserialization_, the Kotlin FHIR library uses the `resourceType` JSON property as the
+[JSON class discriminator](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/json.md#class-discriminator-for-polymorphism).
+This serves as a hint to the `kotlinx.serialization` library so it can dynamically select the
+correct FHIR resource type, or subclass, to instantiate based on the JSON content at runtime.[^5]
+
+[^5]: If the resource type is known at compile time, it is safe to use a more specific type 
+parameter while deserializing, such as `decodeFromString<Patient>(jsonString)`, provided that the
+`Json` object is configured with `ignoreUnknownKeys = true`. The `ignoreUnknownKeys` option is
+critical because with the more specific type parameter the `kotlinx.serialization` library knows the
+specific type to instantiate and no longer uses the `resourceType` JSON property as a
+[JSON class discriminator](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/json.md#class-discriminator-for-polymorphism)
+as is the case when the type parameter `<Resource>` is used. Instead, it will attempt to map the
+`resourceType` JSON property to a regular field in the specified class (e.g., `Patient`). As a
+result, without `ignoreUnknownKeys = true` an error will occur since the target class does not have
+a corresponding field for the `resourceType` property.
+
+However, during _serialization_, it is possible to call the `encodeToString` function without the
+type parameter, or with the type parameter of the specific resource type if it is known at compile
+time, e.g. `encodeToString<Patient>`. But this would be a mistake, since the `kotlinx.serialization`
+library would not treat the serialization as polymorphic in these cases. As a result, it would not
+include the `resourceType` property in the serialized result, generating malformed FHIR JSON.
+Therefore, when serializing FHIR resources, always make sure to include the `<Resource>` type
+parameter like this: <code>encodeToString<b>\<Resource\></b></code>.
 
-    // Use:
-    val resource = json.decodeFromString<Resource>(jsonString)
-    check(resource is Patient)  // True
+```kotlin
+import com.google.fhir.model.r4.Resource
+
+fun main() {
+    // ✅ DO: 
+    val validPatient = json.encodeToString<Resource>(patient)
+    
+    // ⛔ DON'T:
+    // val invalidPatient = json.encodeToString<Patient>(patient)
```

Comment:
Is this happening because of the missing `resourceType` discriminator? I think it will be confusing not to be able to `encodeToString` with `Patient` because `Patient` is indeed a FHIR resource. Can we override the `resourceType` in the FHIR resources to address this problem?

Reply:
I just found out about this: https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-json/kotlinx.serialization.json/-class-discriminator-mode/-a-l-l_-j-s-o-n_-o-b-j-e-c-t-s/

Let me test it.

Reply:
The `ALL_JSON_OBJECTS` option did not work since it would start trying to output a `resourceType` field for data types that inherit from `DataType` and `Element` even though they don't have a `SerialName`...

This is because we set the `classDiscriminator` on the entire module... Together with your other comment, I think this is beginning to point to an approach where we wrap the API a bit more? instead of letting the user to configure the Json?

So... If you're ok with it I can send a seprarate PR to deal with both issues since we need to consider that a bit more carefully.

Reply:
Yeah, I am okay with doing that on a separate PR.